README.md

```diff
@@ -1,3 +1,171 @@
 # npecl design
 
-more info as it is written...
+A work in progress design documentation for the "new pecl" iteration.
+
+Basis:
+
+ * [Discussion doc](https://docs.google.com/document/d/1_N0E9xo3jn9aKrIZHIbTYaY5lXw71BpSO6-it4cRpDo/edit)
+ * Internal PHPF dev meet kick-off to review (minutes?)
+
+## Assumptions
+
+ * npecl is the assumed name, although it is not confirmed
+ * `npecl` is a CLI tool, probably bundled as a PHAR like Composer is
+   * therefore a usage such as `npecl install <thing>` assumes the PHAR exist in `$PATH`, for example in
+     `/usr/bin/npecl`. If not, you would do something like `php npecl.phar install <thing>`.
+
+## npecl itself
+
+ * use `symfony/console` to simplify writing the CLI tool itself
+ * use `composer/composer` to resolve dependencies; it knows how to do all this stuff already, lets not re-invent the
+   wheel - once we have a better idea of what calls we need to make, it may make sense to split some of
+   `composer/composer` out into separate libraries we can consume; although that might be a huge undertaking in itself,
+   so is really just a "strech goal".
+ * There may be multiple versions of php on a system. The version of php invoking `npecl` would be assumed, and its
+   corresponding `phpize` tool would be used.
+ * `npecl` would probably need to be run with `sudo` access (so it can `make install` to root-owned paths). Care should
+   be taken to ensure only the minimal amount is run with elevated privileges, so we don't accidentally give root to a
+   malicious PECL package.
+
+@todo document the CLI optionas
+
+## Extension maintainer: register an npecl package
+
+Create a `composer.json` in your repository, commit:
+
+```json
+{
+    "name": "ext-xdebug",
+    "type": "php-ext-zend",
+    "license": "Xdebug-1.03",
+    "description": "Xdebug is an amazing tool...",
+    "require": {
+        "ext-zlib": "*",
+        "php": ">=7.4.21,<8.4"
+    },
+    "conflict": {
+        "ext-zetalib": "*"
+    },
+    "php-ext": {
+        "priority": 90,
+        "config": {
```

Comment:
@derickr did we reach consensus on what these config options should look like? As I understand it there's two main config options, i.e. 

 - flags, e.g. `--just-a-flag` or `--with-flag` or `--without-flag`
 - options, e.g. `--key=value` (e.g. path to a library)

What about adopting the approach in Pickle?

https://github.com/FriendsOfPHP/pickle/blob/93d76e03b7beb80c22846d2533db12ddfcae57ce/features/pickle/install.feature#L15-L20

Or maybe something else? Seems a bit loose to have "yes" or "no" as the default for "flags" I think...

Reply:
```
"always-prompt": false,
```

That should be the default — if you want to force the prompt (and be an evil extension maintainer), you can set it to `true`. I lean towards not even having this feature.

I am also not sure whether an *option* with a default value is even needed — right now, these are usually used for specifying installation directories of libraries that they require. If **not** given, they do their own heuristics of finding them. If **given**, they then first attempt the given directory.

That would also mean you no longer need the `type` element, as you can detect that by the existence of a default value (a *flag*) or not (an *option*).

I think this might be overcomplicated, considering that I suspect most of the installations are going to be on CI (where  you can't do interactive prompts).

I would only have two modes:

1. `npecl install extname` — would just pass all the default flags and options from the list (skipping the options where there is no default)
2. `npecl install extname -i` — enabling **i**nteractive mode (or if you must, the much longer `--prompt-for-configurations`). `-i` is commonly used, though. In this mode it prompts all questions.

Reply:
OK I think I missed an important point that many extensions will usually build just fine without passing anything to `./configure` - lets simplify much further to start with, and just have:

```json
{
    "php-ext": {
        "priority": 80,
        "configure-options": [
            {
                "name": "enable-xdebug-dev",
                "description": "Enable developer flags"
            },
            {
                "name": "without-xdebug-compression",
                "description": "Disable compression through zlib"
            },
            {
                "name": "some-path-to-something",
                "description": "This should be the path to the thing that is needed."
            }
        ]
    }
}
```

Lets not have an interactive mode (at least to start with), so if you wanted to specify any of the configure options above, it would just be allowed in the CLI options passed to `build` or `install`, some examples:

```bash
$ npecl install xdebug
$ npecl install xdebug --enable-xdebug-dev
$ npecl install xdebug --enable-xdebug-dev --without-xdebug-compression
$ npecl install xdebug --some-path-to-something=/usr/local/lib/something
$ npecl install xdebug --not-a-defined-configuration-option # this would fail
```

We can always improve this later therefore (by using a list of objects), and add new properties here, without breaking BC.